src/mbgl/storage/file_source_request.hpp

```diff
@@ -0,0 +1,31 @@
+#pragma once
+
+#include <mbgl/storage/file_source.hpp>
+#include <mbgl/util/async_request.hpp>
+
+#include <memory>
+#include <functional>
+
+namespace mbgl {
+
+class Mailbox;
+
+class FileSourceRequest : public AsyncRequest {
+public:
+    FileSourceRequest(std::shared_ptr<Mailbox> mailbox);
+    ~FileSourceRequest() final;
+
+    void onResponse(FileSource::Callback&& callback);
+    void onCancel(std::function<void()>&& callback);
+    void setResponse(const Response& res);
+
+private:
+    void runResponseCallback();
+
+    std::shared_ptr<Mailbox> mailbox;
+
+    FileSource::Callback responseCallback = [](Response) {};
+    std::function<void()> cancelCallback = []{};
```

Comment:
Having the cancel callback in here means that we're allocating more space for callbacks that are only used in one implementation.